### PR TITLE
Join stake pool endpoint (part 2 - add sign delegation cert)

### PR DESCRIPTION
**Pull request by paweljakubas**
# Issue Number

#896 

# Overview

- [x] I have adjusted naming as suggested by Johannes in part 1 PR
- [x] I have added signDelegationCert impl
- [x] I have shared create/sign/submit between Tx and DelegationCert 
- [x] I have adjusted and added new integration tests

# Comments

### CODE REVIEW

File: lib/core/src/Cardano/Wallet.hs [hunk 1]
```diff
@@ -112,12 +112,12 @@ module Cardano.Wallet
...
-    , submitCert
-    , ErrCreateCert (..)
-    , ErrSignCert (..)
-    , ErrSubmitCert (..)
+    , createDelegationCert
+    , signDelegationCert
+    , submitDelegationCert
+    , ErrCreateDelegationCert (..)
+    , ErrSignDelegationCert (..)
+    , ErrSubmitDelegationCert (..)
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs]: :+1:

File: lib/core/src/Cardano/Wallet.hs [hunk 2]
```diff
@@ -1021,23 +1024,71 @@ signCert
...
+                    { status = Pending
+                    , direction = Outgoing
+                    , slotId = txSlot
+                    , blockHeight = (currentTip cp) ^. #blockHeight
+                    , amount = Quantity (amtInps - amtOuts)
+                    }
+            let time = slotStartTime
+                    (slotParams (blockchainParameters cp))
+                    txSlot
+            return (undefined, meta, time, undefined)
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs]: **[question]**

What about trying to factor the common parts of this and `signTx` in one or more composable functions? I see arguably at least 2 functions:

- change creation

```hs
            let (changeOuts, s') = flip runState (getState cp) $
                    forM chgs $ \c -> do
                        addr <- state (genChange argGenChange)
                        return $ TxOut addr c
            let allOuts = outs ++ changeOuts
```

- metadata construction

```hs
            let ourCoins (TxOut addr (Coin val)) =
                    if fst (isOurs addr s')
                        then Just (fromIntegral val)
                        else Nothing
            let amtOuts = sum (mapMaybe ourCoins allOuts)
            let amtInps = fromIntegral $
                    sum (getCoin . coin . snd <$> ins)
            let txSlot =
                    (currentTip cp) ^. #slotId
            let meta = TxMeta
                    { status = Pending
                    , direction = Outgoing
                    , slotId = txSlot
                    , blockHeight = (currentTip cp) ^. #blockHeight
                    , amount = Quantity (amtInps - amtOuts)
                    }
            let time = slotStartTime
                    (slotParams (blockchainParameters cp))
                    txSlot
```

Perhaps making them separate function internal to the module would help to identify further commonalities?

**Comment by KtorZ** [comment]
I'll carry on with this PR on top of @Anviking's PR

**Comment by KtorZ** [comment]
@Anviking May you have a first round of review on that one please :pray: 
I will add integration tests in a bit.

File: lib/core/src/Cardano/Wallet.hs (hunk 2 above)
Comment (KtorZ) [reply lib/core/src/Cardano/Wallet.hs]: done.

File: lib/core/src/Cardano/Wallet/Api/Server.hs [hunk 3]
```diff
@@ -820,22 +819,22 @@ joinStakePool ctx spl (ApiT poolId) (ApiT wid) passwd = do
...
 
     let (ApiWalletPassphrase (ApiT pwd)) = passwd
 
     selection <- liftHandler $ withWorkerCtx ctx wid liftE1 $ \wrk ->
-        W.createCert @_ @s @t wrk wid
+        W.selectCoinsForDelegation @_ @s @t wrk wid
 
     (tx, meta, time, wit) <- liftHandler $ withWorkerCtx ctx wid liftE2 $ \wrk ->
-        W.signCert @_ @s @t @k wrk wid () pwd selection poolId
+        W.signDelegation @_ @s @t @k wrk wid () pwd selection poolId
```
Comment (Anviking) [inline lib/core/src/Cardano/Wallet/Api/Server.hs:835]: I'm wondering if we don't have a potential race-condition with `POST /transaction` and `joinStakePool`:

We run:
1. `selectCoinsForDelegation` does the coin selection based on the `availableUtxo`.
2. `signDelegation`
3. `submitTx` — modifies the `availableUtxo` by adding a pending transaction

Step 1 and 3 are run in different DB transactions.

So if we in parallel run:
```
joinStakePool            1           23
joinStakePool                  1       2   3
-------------------------------------------------> time
```
we could end up (attempting to) spend the same funds twice.

Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/Api/Server.hs:835]: Indeed, but that's nothing new :) 
We already have that particular race condition from making two concurrent payments in parallel. Worst that can happen is that we create and sign both transactions, we submit both, and only one get through in the end while the other stay pending forever.

Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/Api/Server.hs:835]: > we could end up spending the same funds twice.

No! The network will prevent that!

Comment (Anviking) [reply lib/core/src/Cardano/Wallet/Api/Server.hs:835]: > No! The network will prevent that!

🙄 Ok, bad wording on my part

**Comment by KtorZ** [comment]
@paweljakubas , okay, so I completed the implementation continuing what you started + doing some refactorings to avoid too much duplication with the standard transactions. 

Also started to enable integration scenarios from the skeleton we started some times ago: 289d9ad44a1a8c2b07af489d2a12eff72467111f

Turned out that this test was not passing. After a bit of investigation, I wrote my findings here:

https://github.com/input-output-hk/jormungandr/issues/1206

I am not sure if we're going to need to work-around that or, if Jörmungandr will get a fix.

File: lib/core/src/Cardano/Wallet.hs [hunk 4]
```diff
@@ -811,49 +870,118 @@ signTx
...
     tl = ctx ^. transactionLayer @t @k
 
+-- | Construct transaction metadata from a current block header and a list
+-- of input and output.
+mkTxMeta
+    :: IsOurs s
+    => BlockchainParameters
+    -> BlockHeader
+    -> s
+    -> [(TxIn, TxOut)]
```
Comment (Anviking) [inline lib/core/src/Cardano/Wallet.hs:963]: (unrelated) It's been nagging me for a while, but we should maybe add proper types for `(TxIn, Coin)` and `(TxIn, TxOut)` sometime

File: lib/core/src/Cardano/Wallet.hs [hunk 5]
```diff
@@ -811,49 +870,118 @@ signTx
...
-                    (slotParams (blockchainParameters cp))
-                    txSlot
-            return (tx, meta, time, wit)
+
+            -- TODO
+            -- Pre-compute this and store it with the wallet static metadata.
+            let rewardAccount = deriveRewardAccount @k xprv pwd
+
+            let hrp = Bech32.unsafeHumanReadablePartFromText "addr"
+            let accountAddr
```
Comment (Anviking) [inline lib/core/src/Cardano/Wallet.hs]: Is importing `Bech32` and knowing how to create the reward `Address` in `core` is temporary? (it is target-specific)

File: lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs [hunk 6]
```diff
@@ -305,6 +308,20 @@ class HardDerivation key => SoftDerivation (key :: Depth -> * -> *) where
...
+deriveRewardAccount
+    :: ( HardDerivation k
+       , Bounded (Index (AddressIndexDerivationType k) 'AddressK)
+       )
+    => k 'RootK XPrv
+    -> Passphrase "encryption"
+    -> k 'AddressK XPrv
+deriveRewardAccount rootPrv pwd =
+    let accPrv = deriveAccountPrivateKey pwd rootPrv minBound
+    in deriveAddressPrivateKey pwd accPrv MutableAccount minBound
```
Comment (Anviking) [inline lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs:324]: 👍

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs [hunk 7]
```diff
@@ -128,11 +134,26 @@ spec = do
...
+            , expectFieldEqual direction Outgoing
+            ]
+
+        -- Wait for the certificate to be inserted
+        eventually $ do
+            let ep = listTxEp w mempty
+            request @[ApiTransaction n] ctx ep Default Empty >>= flip verify
+                [ expectListItemFieldEqual 0 direction Outgoing
+                , expectListItemFieldEqual 0 status InLedger
+                ]
```
Comment (Anviking) [inline lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:156]: 👍 

Are we eventually going to test that rewards accumulate in the reward account too?

Comment (Anviking) [reply lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:156]: Hm, that must depend on #904. Maybe I should pick it up.

Comment (paweljakubas) [reply lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:156]: yes, requesting joining the stake pool is the first step. the next one is to discover it when accommodated in blockchain

File: lib/core/src/Cardano/Wallet.hs (hunk 5 above)
Comment (KtorZ) [reply lib/core/src/Cardano/Wallet.hs]: Screenshot from 2019-11-27 19-17-39

It is, as indicated by this big fat "TEMPORARY" in the corresponding commit :sweat_smile:

**Review by KtorZ — APPROVED** [review]
@paweljakubas 

- Rebased on top of `KtorZ/896/delegation-addresses-only`
- Removed the temporary hack
- Test now pass as expected :tada:

**Pull request merged**